[Async object upload 2] Add struct chunkUploader

gargnitingoogle opened the pull request:
Adds a new struct chunkUploader which implements the ChunkUploader interface.

### Description
A chunkUploader is used to write content to a new GCS object asynchronously.

This is followed up in and [Async object upload 3](https://github.com/GoogleCloudPlatform/gcsfuse/pull/1417).

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - All existing unit tests passed. New tests needed for chunkUploader.
3. Integration tests - NA

gargnitingoogle commented on internal/storage/streamed_object_writer_test.go:
```diff
@@ -0,0 +1,38 @@
+// Tests
+////////////////////////////////////////////////////////////////////////
+
+func (t *StreamedObjectWriterTest) DummyWriter() {
```
Done

gargnitingoogle commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	Writing
+	// Last write failed.
+	WriteError
+	// Close has been failed.
```
Fix typo

gargnitingoogle commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+
+func (sow StreamedChunkUploader) ChunkSize() (int, error) {
+	if nil == sow.wc {
+		return 0, fmt.Errorf("writer not initiated")
```
```suggestion
		return 0, fmt.Errorf("writer not initialized")
```

gargnitingoogle commented on internal/storage/streamed_chunk_uploader_test.go:
```diff
@@ -0,0 +1,329 @@
+	AssertEq(contentSize, uploader.BytesUploadInitiatedSoFar())
+}
+
+func (t *StreamedChunkUploaderTest) TestUploadMultipleMultichunksUploads() {
```
```suggestion
func (t *StreamedChunkUploaderTest) TestUploadMultipleMultichunkUploads() {
```

gargnitingoogle commented on internal/storage/streamed_chunk_uploader_test.go:
```diff
@@ -0,0 +1,329 @@
+		AssertEq(0, uploader.BytesUploadInitiatedSoFar())
+		AssertEq(0, uploader.BytesUploadedSoFar())
+		chunkSize, err := uploader.ChunkSize()
+		AssertEq(nil, err)
```
Duplicate line. remove!

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+// It also stores state of how many bytes have been sent to it
+// to write, how many have been successfully written by it so far,
+// and the current state of the uploader.
+type StreamedChunkUploader struct {
```
Why should this struct be public. As the responsibility of this struct it to implement the chunkUploader interface, the init method should just return the interface type.

gargnitingoogle replied:
Made private.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+// It also stores state of how many bytes have been sent to it
+// to write, how many have been successfully written by it so far,
+// and the current state of the uploader.
+type StreamedChunkUploader struct {
```
why is this called StreamedChunkUpload and not chunkUploader simply?

gargnitingoogle replied:
Renamed to chunkUploader.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+// to write, how many have been successfully written by it so far,
+// and the current state of the uploader.
+type StreamedChunkUploader struct {
+	// Implements the following interfaces.
```
If the struct implements all the interface methods, it is assumed that struct implements the interface. no need of this line.

gargnitingoogle replied:
@vadlakondaswetha Do you mean just the comment or the interface declaration on the next line as well is not needed? 

I know the interface declaration isn't necessary, but I've made it explicit to make it easier for the reader, e.g. in https://github.com/GoogleCloudPlatform/gcsfuse/blob/66b8100310ff701a7f4fba5f801615c3480228dc/internal/gcsx/content_type_bucket.go#L32 .

vadlakondaswetha replied:
The reason for adding inside content_type_bucket struct is to store the reference of original implementation and call it in other methods:

https://github.com/GoogleCloudPlatform/gcsfuse/blob/66b8100310ff701a7f4fba5f801615c3480228dc/internal/gcsx/content_type_bucket.go#L28
https://github.com/GoogleCloudPlatform/gcsfuse/blob/66b8100310ff701a7f4fba5f801615c3480228dc/internal/gcsx/content_type_bucket.go#L44

 I dont think there is any other definition which uploadChunker wants to call.

gargnitingoogle replied:
Ok. Removed.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// Chuck size for resumable upload is default i.e. 16MB.
+	wc := obj.NewWriter(ctx)
+	// NewWriter never returns nil, so no nil-check here.
```
nit: add comments on top of line not below.

gargnitingoogle replied:
The null check was needed later, so I added the comment later, to keep the flow for the reader. Have moved before the line now.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	sow.userProgressFunc = progressFunc
+
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// Chuck size for resumable upload is default i.e. 16MB.
```
nit: chuck -> chunk

gargnitingoogle replied:
This comment line was outdated, so removed it.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+func (sow StreamedChunkUploader) ChunkSize() (int, error) {
+	if nil == sow.wc {
```
nit: sow.wc == nil

gargnitingoogle replied:
Done.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+func (sow StreamedChunkUploader) BytesUploadedSoFar() int64 {
+	return sow.totalWriteSucceededSoFar
```
why are these methods public, how will the caller use them?
I assume caller should use chunkUploader interface to upload and query the status.

vadlakondaswetha replied:
Are these methods only required in unit tests. If yes you should be able to access private methods since both the files are in same package. refer to bucket_handle_test

gargnitingoogle replied:
`BytesUploadedSoFar()` is declared inthe ChunkUploader interface, so I've kept it public, and have made the non-interface functions private.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+
+func (sow *StreamedChunkUploader) readyToWrite() bool {
+	switch sow.state {
+	case Initialized:
```
why dont you combine everything as below:

switch sow.state {
	case
		Initialized,
		Waiting,
		Writing:
		return true
	}
	return false

gargnitingoogle replied:
Done.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+func (sow *StreamedChunkUploader) readyToClose() bool {
+	switch sow.state {
+	case Initialized:
+		fallthrough
```
same here also.

gargnitingoogle replied:
Done.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	}
+}
+
+// Write uploads the given content
```
the comment is not clear. Please rephrase in simpler terms

gargnitingoogle replied:
Done. Simplified to one sentence.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+
+	n, err := io.Copy(sow.wc, contents)
+	sow.totalWriteInitiatedSoFar += n
+	if n == 0 || err == io.EOF {
```
why are you converting n==0 to io.EOF? just make it a nop when no data is passed?

gargnitingoogle replied:
Done. Removed the `n==0` case.

gargnitingoogle replied:
Yeah, I was wrong earlier, io.EOF is never returned from io.Copy in successful case unless size is specified. Since we're specifying size here, we can't get io.EOF. So, I've removed any special handling for io.EOF.

As for n==0, I've treated that a NOP now.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	case Waiting:
+		fallthrough
+	case Writing:
+		fallthrough
```
what happens if the chunk upload is in progress and we call the close. Will the chunk gets uploaded successfully before finalizing the object?
If yes can you document it on the close method of chunkUploader interface.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	return nil
+}
+
+func (sow *StreamedChunkUploader) Close() (o *gcs.Object, err error) {
```
the purpose of assigning variable names to the object is to set the values using these names.


For eg: 
o =storageutil.ObjectAttrsToBucketObject(attrs)
return

If you are not doing that ever in the code, dont use variable names in the method definition.

gargnitingoogle replied:
Removed unused names.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+		return nil, fmt.Errorf("writeChunkSize <= 0 not supported")
+	}
+
+	// raw initialization
```
nit: Please start all comments with capital letter and end with fullstop;.

gargnitingoogle replied:
Done.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	// Chuck size for resumable upload is default i.e. 16MB.
+	wc := obj.NewWriter(ctx)
+	// NewWriter never returns nil, so no nil-check here.
+
```
nit: remove new line here.

gargnitingoogle replied:
Done.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	if n == 0 || err == io.EOF {
+		// for io.EOF, don't change/append the error (message) as it makes it difficult to identify EOF for the caller.
+		sow.state = Waiting
+		return io.EOF
```
what happens if you try to write data < chunksize, will the writer wait till the next set of data comes?

gargnitingoogle replied:
No, it doesn't wait, it just returns with io.EOF, which is a harmless error.

vadlakondaswetha replied:
When does the upload happen of that data happens? when we write the next chunk or call close?

gargnitingoogle replied:
It happens during the next upload call(s) whenever a full chunk is available.

vadlakondaswetha commented on internal/storage/streamed_chunk_uploader.go:
```diff
@@ -0,0 +1,206 @@
+	}
+
+	// We can't use defer to close the writer, because we need to close the
+	// writer successfully before calling Attrs() method of writer.
```
this comment doesn't make any sense. this method is triggered to finalize the upload and it has to be done synchronously.

gargnitingoogle replied:
Removed the unnecessary and confusing comment.

vadlakondaswetha reviewed (commented):
Unit tests are not reviewed.

gargnitingoogle commented:
> Unit tests are not reviewed.

Understood.

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+		AssertGe(numChunks, numCallbacks)
+		AssertGe(contentSize, numBytesUploaded)
+	}
+	AssertEq(numBytesUploaded, uploader.BytesUploadedSoFar())
```
Isn't this wrong?  you are trying to compare with values generated from the test progress function instead of absolute values. Basically numBytesUploaded and uploader.BytesUploadedSoFar() are generated from the same function

gargnitingoogle replied:
Fixed.

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	AssertEq(contentSize, uploader.bytesUploadInitiatedSoFar())
+}
+
+func (t *ChunkUploaderTest) TestUploadMultipleMultichunkUploads() {
```
Same comments for all test methods 
1. formatting
2. comparision part
3. adding comments.

gargnitingoogle replied:
Done.

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	AssertEq(contentSize, uploader.bytesUploadInitiatedSoFar())
+}
+
+func (t *ChunkUploaderTest) TestUploadMultipleMultichunkUploads() {
```
What about testcase for TEstMultipleSingleChunkUploads ?

gargnitingoogle replied:
I've added a single generic test for multiple uploads of mixed sizes, which covers all cases.

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+
+func (t *ChunkUploaderTest) TestCloseWithoutWrite() {
+	obj := t.obj
+
```
nit: remove new line

gargnitingoogle replied:
Done

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	o, err := uploader.Close()
```
nit: add new line after uploader.Close().

gargnitingoogle replied:
Done

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	AssertEq(contentSize, o.Size)
+}
+
+func (t *ChunkUploaderTest) TestCloseWithMultipleSmallUploads() {
```
can you define what does small means in the test method itself.

gargnitingoogle replied:
Small stands for where content-size < chunkSize.

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	AssertEq(contentSize, o.Size)
+}
+
+func (t *ChunkUploaderTest) TestCloseWithMultipleBigUploads() {
```
what does big mean? add it in the test method name itself.

gargnitingoogle replied:
Big stands for where content-size > chunkSize.

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+		"iy3rh34r489y89j309",
+		"ie32hr83hr43rt9y8do93j9",
+		"i9r38ry2u9j9orj3"}
+	var contentSize int
```
Negative test cases are missing
when empty contents are passed to uploadChunkAsync?
when io.EOF is returned
Can you trigger an error scenario - invalid io.Reader and what happens when you call close?

gargnitingoogle replied:
> when empty contents are passed to uploadChunkAsync?

Added test for passing empty content to uploadChunkAsync.

> when io.EOF is returned

io.EOF is never returned, so removed that check from implementation.

> invalid io.Reader and what happens when you call close?

This needs work.

gargnitingoogle replied:
Added test for failed uploader and another test for showing its sucessful closure.

vadlakondaswetha commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+		contentSize += len(content)
+	}
+
+	o, err := uploader.Close()
```
Please validate the uploader state after close in all test cases

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,186 @@
+// for resumable upload.
+//
+// It also stores the current state of the uploader.
+type chunkUploader struct {
```
Added mutex and locks on all public calls to chunkUplader.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	for _, input := range inputs {
+		uploader, err := NewChunkUploader(input.ctx, input.obj, input.req, input.writeChunkSize, input.progressFunc)
+		AssertEq(nil, uploader)
+		AssertNe(nil, err)
```
Added check on err.Error()

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,186 @@
+
+// NewChunkUploader creates a new instace of StreamedChunkUploader,
+// for the given inputs.
+func NewChunkUploader(ctx context.Context, obj *storage.ObjectHandle, req *gcs.CreateObjectRequest, writeChunkSize int, progressFunc func(int64)) (*chunkUploader, error) {
```
Done.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,186 @@
+	return uploader.totalWriteInitiatedSoFar
+}
+
+func (sow chunkUploader) BytesUploadedSoFar() int64 {
```
Done.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,186 @@
+	state UploaderState
+}
+
+// NewChunkUploader creates a new instace of StreamedChunkUploader,
```
Done.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+func (t *ChunkUploaderTest) SetUp(ti *TestInfo) {
+	t.fakeStorage = NewFakeStorage()
+	t.storageHandle = t.fakeStorage.CreateStorageHandle()
+	t.bucketHandle = t.storageHandle.BucketHandle(TestBucketName, "")
```
Removed both storageHandle and bucketHandle from struct ChunkUploaderTest .

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	AssertEq(Writing, uploader.state)
+	AssertEq(contentSize, uploader.bytesUploadInitiatedSoFar())
+
+	// time.Sleep(100 * time.Millisecond)
```
No, removed. I've synchronized based on a channel now.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+		AssertGt(1, numCallbacks)
+		AssertGt(contentSize, numBytesSuccessfullyUploaded)
+	}
+	AssertEq(numBytesSuccessfullyUploaded, uploader.BytesUploadedSoFar())
```
Uploaded the code now. I've synchronized based on a channel now.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	AssertEq(Writing, uploader.state)
+
+	// time.Sleep(1000 * time.Millisecond)
+	if numCallbacks > 0 {
```
Uploaded the code now. I've synchronized based on a channel now.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+
+	AssertEq(Writing, uploader.state)
+
+	// time.Sleep(1000 * time.Millisecond)
```
Removed.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+
+	// time.Sleep(100 * time.Millisecond)
+	if numCallbacks > 0 {
+		AssertGt(1, numCallbacks)
```
Fixed.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+
+	// time.Sleep(100 * time.Millisecond)
+	if numCallbacks > 0 {
+		AssertGt(1, numCallbacks)
```
Fixed.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,330 @@
+	}
+}
+
+func (t *ChunkUploaderTest) TestUploadSingleChunkSingleUpload() {
```
Fixed.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,592 @@
+	ctx := context.Background()
+
+	// cleanup
+	defer func() {
```
No need for func() here.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,592 @@
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	// test payload
```
Move this comment down.

ashmeenkaur commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,202 @@
+		return nil, fmt.Errorf("ctx is nil")
+	}
+	if obj == nil || req == nil {
+		return nil, fmt.Errorf("nil obj or req")
```
NIT: nil ObjectHandle or CreateObjectRequest

gargnitingoogle replied:
Done.

ashmeenkaur commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	if obj.ObjectName() != req.Name {
+		return nil, fmt.Errorf("names of passed object-handle and createObjectRequest.Name don't match: object-handle=%s createObjectRequest.Name=%s", obj.ObjectName(), req.Name)
```
let's use camelcase for ObjectHandle as well

gargnitingoogle replied:
Done.

ashmeenkaur replied:
there is one more object-handle

gargnitingoogle replied:
Removed.

ashmeenkaur reviewed (commented):
added some comments. Need to review unit tests as well.

ashmeenkaur commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,588 @@
+package storage
```
there are a lot of common assertions in this test file, can we move them to a common helper function?

gargnitingoogle replied:
Clubbed all the common assertions around the new/upload/close calls into helper functions. 

I am not sure it's good idea to try to optimize/share the test logic easily without making it harder to change them later if needed.

gargnitingoogle replied:
@ashmeenkaur reminder.

ashmeenkaur replied:
Thank you!

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,202 @@
+package storage
```
Yes, good idea.

gargnitingoogle replied:
Done.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,202 @@
+
+func (uploader *chunkUploader) readyToWrite() bool {
+	switch uploader.state {
+	case Initialized, Waiting, Writing:
```
@ashmeenkaur  Yes, the default state of a non-nil chunkUploader will be Initialized. I didn't understand what you meant by 'shall we change it' ?

ashmeenkaur replied:
I was concerned if someone can run into a case where they create a zero value instance of chunk with Initialized state but upon thinking more, I believe we can never run into that case considering the struct is private and the only way to create an instance is via NewChunkUploader.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,202 @@
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = writeChunkSize
+	wc.ProgressFunc = func(n int64) {
+		uploader.totalWriteSucceededSoFar = n
```
This callback doesn't indicate if there is any more data upload in progress or not, so not a good idea to set it here. Plus changing it here will require a mutex lock which can slow things down unnecessary, or even a deadlock.

gargnitingoogle replied:
I have now removed the `Waiting` state altogether as it didn't make much sense after more thought. So, this is now a moot point.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,202 @@
+// This waits (using mutex) until the completion of
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+func (uploader *chunkUploader) UploadChunkAsync(contents io.Reader) error {
```
> This method has no logic related to async upload

The word async in function name indicates that this function doesn't block on the upload to complete. This is a common convention.

> I am a little concerned about retries and timeouts for this call since it doesn't take context. Context aware reader is one option but I was wondering if go client has something related?

Let's take this separately in another PR. This needs discussion and possibly interface change.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,202 @@
+	uploader.mu.Lock()
+	defer uploader.mu.Unlock()
+
+	if !uploader.readyToWrite() || (uploader.wc == nil) {
```
This checke was redundant. Removed now.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,202 @@
+	return &uploader, nil
+}
+
+func (uploader *chunkUploader) chunkSize() (int, error) {
```
Was redundant. Removed.

gargnitingoogle commented:
@ashmeenkaur I addressed your comments. There are follow-ups on a few of them, PTAL.

ashmeenkaur commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,214 @@
+	// StorageCopyTimeoutDuration is how long the chunkUploader waits for the
+	// copy call to the uploader to return. If it takes longer, the
+	// upload call times out.
+	StorageCopyTimeoutDuration time.Duration = time.Second
```
As discussed offline, let's remove the timeout logic from asyncChunkUpload and this const

gargnitingoogle replied:
ok

gargnitingoogle replied:
Done.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,214 @@
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+func (uploader *chunkUploader) BytesUploadedSoFar() int64 {
+	return atomic.LoadInt64(&uploader.totalWriteSucceededSoFar)
```
`atomic.LoadInt64` is basically for thread-safety. It's needed because this variable is updated in the callback (different goroutine) and is read in the goroutine (calling the `uploader.Close()` or `uploader.Upload...`).

I didn't know that there was a clearner alternative to it. I'll switch to `atomic.Int64.Load()` .

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,214 @@
+// It also stores the current state of the uploader.
+type chunkUploader struct {
+	// Internal objects for business logic.
+	wc         *storage.Writer
```
No specific reason, the seed code for this struct was taken by `bucket_handle.CreateObject()`, and also the names. I'll change it.

ashmeenkaur commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,463 @@
+
+func (t *ChunkUploaderTest) testFailedClose(uploader *chunkUploader) {
+	_, err := uploader.Close()
+	AssertNe(nil, err)
```
Should we assert on the error string as well?

gargnitingoogle replied:
Added assert on error substring.

ashmeenkaur reviewed (commented):
Minor comments, LGTM otherwise.
Also noticed some error statements going beyond 80 characters and comments < 80 characters. Can we make it consistent?

gargnitingoogle reviewed (commented):
Addressed comments.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,463 @@
+	}
+}
+
+func (t *ChunkUploaderTest) TestUploadMultipleUploads() {
```
Done. Added a comment explaining the flow, and also added a few functions to simplify the code flow.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,463 @@
+		}
+	}
+}
+
```
Also added another test TestMultipleHomogeneousUploads .

ashmeenkaur replied:
I see the timeout is 100ms. Is this intended?

gargnitingoogle replied:
Yes, I increased the rest of timeouts to 10ms to avoid some flakiness with 1ms. So, increased this one to 100ms.

gargnitingoogle commented:
@ashmeenkaur all comments addressed. PTAL.

gargnitingoogle commented:
@sethiay @raj-prince please review and approve.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,193 @@
+
+	// Attributes for providing updates to user.
+	totalWriteInitiatedSoFar int64
+	// Total number of bytes successfully written by this writer so far.
```
Yes. Changed all places.

sethiay replied:
let's please change the code comments also.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,193 @@
+// This waits (using mutex) until the completion of
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+// The call times out if it is taking longer than StorageCopyTimeoutDuration.
```
Stale comment line.. Removed.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,193 @@
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+// The call times out if it is taking longer than StorageCopyTimeoutDuration.
+func (uploader *chunkUploader) UploadChunkAsync(contents io.Reader) error {
```
No, the timeout will be in the caller.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,193 @@
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+// The call times out if it is taking longer than StorageCopyTimeoutDuration.
+func (uploader *chunkUploader) UploadChunkAsync(contents io.Reader) error {
```
The function is asynchronous as it queues up the GCS uploads and returns right after queueing up the last chunk provided in the content, rather than blocking on that chunk-upload finishing. When the uploaded content is a single chunk (which is the main use case), then it just queues up the gcs upload and returns. There will not be much value to calling this in a goroutine if we stick to single-chunk uploads.

sethiay replied:
When the buffer size is greater than 50 MiB, chunk uploader will get `contents` which has more than 1 chunk size of data. So in that case, this will be a blocking call for caller. Should we do something to handle this ?

gargnitingoogle replied:
@sethiay As discussed offline, if the uploader has to be used for uploading multiple chunks, then this function should be called in a goroutine on the caller side. 
I'll add the context for timing out the call in a goroutine as you suggested though.

gargnitingoogle replied:
Done.

gargnitingoogle commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,193 @@
+		return fmt.Errorf("writer not ready to write: object: %s, status = %v, writer: %v", uploader.objectName, uploader.state, uploader.writer)
+	}
+
+	n, err := io.Copy(uploader.writer, contents)
```
No, as per the documentation of io.Copy, io.EOF is not considered an error and is just retuned as `err: nil`.

gargnitingoogle commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,497 @@
+	}
+}
+
+func (t *ChunkUploaderTest) CreateWithProperInputs() {
```
Yes. It's a test. Renamed.

gargnitingoogle commented on internal/storage/gcs/bucket.go:
```diff
@@ -31,9 +31,6 @@ type ChunkUploader interface {
 	// Progress should be tracked using the progress func
 	// passed during ChunkUploader instance creation.
 	//
-	// io.EOF is an expected error in case the reader has less
```
@sethiay  You mean what the comment says isn't true for GCS? If so, I've removed that comment.

sethiay replied:
Sorry, I meant "removing the comment" isn't true for real GCS (or the comment which was there is correct for real GCS).

gargnitingoogle replied:
Resolving as discussed offline as this is a stale comment removed.

gargnitingoogle commented:
@sethiay addressed all your comments.

Changes:
1. Passed context into chunkUploader.{Upload(), Close()} for allowing them to be cancelled by the caller. Added additional unit tests for them.
2. Renamed ChunkUploader.UploadChunkAsync() to Upload to generalize it.

cc @ashmeenkaur PTAL.

sethiay commented on internal/storage/chunk_uploader.go:
```diff
@@ -0,0 +1,233 @@
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("uploader.Upload() timed out")
```
nit: whether it's timed out or cancelled because of other reasons, we don't know. So, can we use ctx.Err() here instead of hard coding timed out?

gargnitingoogle replied:
Done.

sethiay commented on internal/storage/chunk_uploader_test.go:
```diff
@@ -0,0 +1,523 @@
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond)
+	defer cancel()
+	err := uploader.Upload(ctx, strings.NewReader(content))
+	AssertNe(nil, err)
```
please also assert with error message.

gargnitingoogle replied:
Done.
